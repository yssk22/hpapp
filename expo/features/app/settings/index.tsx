import { logLogin, logLogout } from '@hpapp/system/firebase';
import * as logging from '@hpapp/system/logging';
import { useCallback, useEffect, useMemo, useState } from 'react';
import { Platform } from 'react-native';

import SettingsAppConfig, { AppConfig, SettingsAppConfigDefault } from './internals/SettingsAppConfig';
import SettingsCurrentUser, { CurrentUser } from './internals/SettingsCurrentUser';
import SettingsList from './internals/SettingsList';
import Provider, { useSettings } from './internals/SettingsProvider';
import SettingsUPFCConfig, { UPFCConfig } from './internals/SettingsUPFCConfig';
import SettingsUserConfig, { UserConfig, SettingsUserConfigDefault } from './internals/SettingsUserConfig';

export type SettingsProviderProps = {
  appConfig?: AppConfig;
  userConfig?: UserConfig;
  currentUser?: CurrentUser;
  upfcConfig?: UPFCConfig;
};

/**
 * A provider to setup the settings context
 */
export function SettingsProvider({
  appConfig,
  userConfig,
  currentUser,
  upfcConfig,
  children
}: SettingsProviderProps & {
  children: React.ReactNode;
}) {
  const [ready, setReady] = useState(false);
  useEffect(() => {
    (async () => {
      if (Platform.OS === 'web') {
        // we expect web is used only for storybook so we set the user token here from the environment variable and enforce the logged in user
        const accessToken = process.env['HPAPP_USER_TOKEN_FOR_JEST'];
        if (accessToken === undefined) {
          throw new Error('HPAPP_USER_TOKEN_FOR_JEST is not set for web');
        }
        await SettingsCurrentUser.save({
          id: '1234',
          username: 'yssk22',
          accessToken
        });
      } else {
        currentUser && (await SettingsCurrentUser.save(currentUser));
        userConfig && (await SettingsUserConfig.save(userConfig));
        appConfig && (await SettingsAppConfig.save(appConfig));
        upfcConfig && (await SettingsUPFCConfig.save(upfcConfig));
      }
      setReady(true);
    })();
  }, []);
  if (!ready) {
    return null;
  }
  return <Provider settings={SettingsList}>{children}</Provider>;
}

/**
 * Get the application configuration
 */
export function useAppConfig() {
  const [appConfig] = useSettings(SettingsAppConfig);
  return useMemo(() => {
    return {
      ...SettingsAppConfigDefault,
      ...appConfig
    };
  }, [appConfig]);
}

export { SettingsAppConfigDefault };

/**
 * Get the updator for AppConfig
 */
export function useAppConfigUpdator() {
  const [, updator] = useSettings(SettingsAppConfig);
  return updator;
}

/**
 * Get the user configuration
 */
export function useUserConfig() {
  const [userConfig] = useSettings(SettingsUserConfig);
  return userConfig;
}

/**
 * Get the updator for user configuration
 */
export function useUserConfigUpdator() {
  const [, updator] = useSettings(SettingsUserConfig);
  return updator;
}

/**
 * Get the UPFC configuration
 */
export function useUPFCConfig() {
  const [upfcConfig] = useSettings(SettingsUPFCConfig);
  // migrate from legacy helloproject credentials to new one to support mline
  return useMemo(() => {
    if (upfcConfig) {
      return {
        ...upfcConfig,
        username: undefined,
        password: undefined,
        // eslint-disable-next-line deprecation/deprecation
        hpUsername: upfcConfig?.username ?? upfcConfig?.hpUsername,
        // eslint-disable-next-line deprecation/deprecation
        hpPassword: upfcConfig?.password ?? upfcConfig?.hpPassword
      };
    }
    return undefined;
  }, [upfcConfig]);
}

/**
 * Get the updator for UPFC configuration
 */
export function useUPFCConfigUpdator() {
  const [, updator] = useSettings(SettingsUPFCConfig);
  return updator;
}

/**
 * Get the current user
 */
export function useCurrentUser() {
  const [currentUser] = useSettings(SettingsCurrentUser);
  return currentUser;
}

export type User = NonNullable<ReturnType<typeof useCurrentUser>>;

/**
 * Get the updator for current user
 */
export function useCurrentUserUpdator() {
  const [currentUser, updator] = useSettings(SettingsCurrentUser);
  return useCallback(
    (update: CurrentUser | null) => {
      if (currentUser) {
        if (update == null) {
          logging.Info('features.auth.logout', 'logout', {
            userid: currentUser.id
          });
          logLogout();
        }
      } else {
        if (update) {
          logging.Info('features.auth.login', 'login', {
            userid: update.id
          });
          // TODO: we should have a way to know the login method
          logLogin('default', update.id);
        }
      }
      return updator(update);
    },
    [updator]
  );
}

export { SettingsUserConfigDefault };
