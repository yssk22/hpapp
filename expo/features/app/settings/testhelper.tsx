import { act, render } from '@testing-library/react-native';

import { SettingsProvider, SettingsProviderProps } from './';
import SettingsList from './internals/SettingsList';

/**
 * Clear all settings
 */
export async function clearSettings() {
  await Promise.all(
    SettingsList.map((settings) => {
      settings.clear();
    })
  );
}

export async function renderSettingsProvider(children: React.ReactNode, props: SettingsProviderProps = {}) {
  await clearSettings();
  const content = render(<SettingsProvider {...props}>{children}</SettingsProvider>);
  await act(async () => {});
  return content;
}
