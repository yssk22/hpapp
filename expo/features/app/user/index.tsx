import { HPArtist, HPMember, HPFollowType, HPFollow } from './internals/HelloProject';
import Me from './internals/Me';
import UserRoot, { UserRootProps } from './internals/UserRoot';
import UserServiceProvider, {
  useArtist,
  useArtistList,
  useArtistMap,
  useMember,
  useMemberList,
  useMemberMap,
  useFollowingArtistList,
  useFollowingMemberList,
  useMe,
  useReloadUserContext
} from './internals/UserServiceProvider';

export {
  HPArtist,
  HPMember,
  HPFollow,
  HPFollowType,
  Me,
  UserRoot,
  UserServiceProvider,
  UserRootProps,
  useArtist,
  useArtistMap,
  useArtistList,
  useMember,
  useMemberList,
  useMemberMap,
  useFollowingArtistList,
  useFollowingMemberList,
  useMe,
  useReloadUserContext
};
