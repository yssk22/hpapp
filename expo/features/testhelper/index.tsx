import {
  mockUseNavigation,
  renderUserScreen,
  renderUserComponent,
  renderGuestComponent,
  waitForUserRootRerendered
} from './internals/jest';
import { ScreenDecorator } from './internals/storybook';

export {
  mockUseNavigation,
  renderUserScreen,
  renderUserComponent,
  renderGuestComponent,
  waitForUserRootRerendered,
  ScreenDecorator
};
