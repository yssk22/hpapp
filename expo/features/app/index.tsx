import { maybeDev } from '@hpapp/foundation/environment';
import { LocalMediaManagerProvider } from '@hpapp/system/media';

import AppRelayProvider from './internals/AppRelayProvider';
import { default as Root, AppRootProps as RootProps } from './internals/AppRoot';
import AppUpdateBanner from './internals/AppUpdateBanner';
import { ThemeProvider, ThemeProviderProps } from './theme';

export type AppRootProps = RootProps & ThemeProviderProps;

export function AppRoot({ screens, ...props }: AppRootProps) {
  return (
    <ThemeProvider {...props}>
      <AppRelayProvider>
        <LocalMediaManagerProvider name={maybeDev() ? 'hellofanapp-dev' : 'hellofanapp'}>
          <AppUpdateBanner />
          <Root screens={screens} />
        </LocalMediaManagerProvider>
      </AppRelayProvider>
    </ThemeProvider>
  );
}
