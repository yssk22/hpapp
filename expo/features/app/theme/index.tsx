import { SettingsProvider, SettingsProviderProps } from '@hpapp/features/app/settings';
import { createTheme, calcYIQ, rgbToColor, colorToRGB } from '@hpapp/system/theme/helper';
import { ThemeColorScheme, ThemeColorKey } from '@hpapp/system/theme/types';
import { useTheme } from '@rneui/themed';
import React, { useCallback } from 'react';

import Provider from './internals/ThemeProvider';

export type ThemeProviderProps = SettingsProviderProps;

/**
 * provides the theme context. It has to be under SettingsProvider.
 */
export function ThemeProvider({ children, ...props }: { children: React.ReactNode } & SettingsProviderProps) {
  return (
    <SettingsProvider {...props}>
      <Provider>{children}</Provider>
    </SettingsProvider>
  );
}

const yiqThreshod = 64;

/**
 * get the color and contrast color from the color scheme
 * @param scheme
 * @returns
 */
export function useThemeColor(scheme: ThemeColorScheme) {
  const { theme } = useTheme();
  const color = theme.colors[scheme];
  const def = rgbToColor(color);
  if (def) {
    return [color, def.contrastColor];
  }
  const yiq = calcYIQ(color);
  return [color, yiq > yiqThreshod ? theme.colors.white : theme.colors.black];
}

/**
 * get the color for the skelton
 * @param scheme
 * @param percent
 * @returns
 */
export function useThemeSkeltonColor(scheme: ThemeColorScheme, percent = 0.8) {
  const { theme } = useTheme();
  const color = theme.colors[scheme];
  const [r, g, b] = colorToRGB(color);
  const r1 = Math.min(255, Math.floor(r + (255 - r) * percent));
  const g1 = Math.min(255, Math.floor(g + (255 - g) * percent));
  const b1 = Math.min(255, Math.floor(b + (255 - b) * percent));
  // convert r1, g1, b1 to hex
  return `#${r1.toString(16).padStart(2, '0')}${g1.toString(16).padStart(2, '0')}${b1.toString(16).padStart(2, '0')}`;
}

/**
 * Get the updator function for the theme
 * @returns a function to update the theme
 */
export function useThemeUpdator() {
  const theme = useTheme();
  return useCallback(
    () =>
      function (primaryColor: ThemeColorKey, secondaryColor: ThemeColorKey, bgColor: ThemeColorKey) {
        theme.updateTheme(createTheme(primaryColor, secondaryColor, bgColor));
      },
    [theme.updateTheme]
  );
}
