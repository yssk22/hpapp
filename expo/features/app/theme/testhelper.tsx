import { SettingsProviderProps } from '@hpapp/features/app/settings';
import { renderSettingsProvider } from '@hpapp/features/app/settings/testhelper';

import { ThemeProvider } from './';

export async function renderThemeProvider(children: React.ReactNode, props: SettingsProviderProps = {}) {
  return renderSettingsProvider(<ThemeProvider>{children}</ThemeProvider>, props);
}
