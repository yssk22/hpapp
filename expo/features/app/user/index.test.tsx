import { renderUserComponent } from '@hpapp/features/app/testhelper';
import { Text } from 'react-native';

import { useHelloProject, useMe } from './';

describe('user', () => {
  test('useHelloProject', async () => {
    function TestComponent() {
      const hp = useHelloProject();
      const artist = hp.useArtist('1'); // morningmusume
      const member = hp.useMember('8589934592'); // mizuki_fukumura
      const artists = hp.useArtists();
      const members = hp.useMembers();
      return (
        <>
          <Text testID="user.test.useArtist">{artist!.key}</Text>
          <Text testID="user.test.useMember">{member!.key}</Text>
          <Text testID="user.test.useArtists">
            {artists
              .map((a) => {
                return a.key;
              })
              .join('\n')}
          </Text>
          <Text testID="user.test.useMembers">
            {members
              .map((m) => {
                return m.key;
              })
              .join('\n')}
          </Text>
        </>
      );
      // return <Text id="test">morningmusume</Text>;
    }
    // // test the useHelloProject function
    const content = await renderUserComponent(<TestComponent />);
    expect(content.getByTestId('user.test.useArtist').props.children).toBe('morningmusume');
    expect(content.getByTestId('user.test.useMember').props.children).toBe('mizuki_fukumura');
    expect(content).toMatchSnapshot();
  });

  test('should return Me object', async () => {
    function TestComponent() {
      const me = useMe();
      return (
        <>
          <Text testID="user.test.username">{me.username}</Text>
        </>
      );
    }
    const content = await renderUserComponent(<TestComponent />);
    expect(content.getByTestId('user.test.username').props.children).toBe('yssk22');
    expect(content).toMatchSnapshot();
  });
});
