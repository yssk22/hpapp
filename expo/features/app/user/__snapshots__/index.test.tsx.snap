// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user should return Me object 1`] = `
<View
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <RNSScreenStack
    style={
      {
        "flex": 1,
      }
    }
  >
    <RNSScreen
      collapsable={false}
      fullScreenSwipeShadowEnabled={false}
      gestureResponseDistance={
        {
          "bottom": -1,
          "end": -1,
          "start": -1,
          "top": -1,
        }
      }
      hasLargeHeader={false}
      nativeBackButtonDismissalEnabled={false}
      onAppear={[Function]}
      onDisappear={[Function]}
      onDismissed={[Function]}
      onGestureCancel={[Function]}
      onHeaderBackButtonClicked={[Function]}
      onNativeDismissCancelled={[Function]}
      onTransitionProgress={[Function]}
      onWillAppear={[Function]}
      onWillDisappear={[Function]}
      replaceAnimation="push"
      sheetAllowedDetents="large"
      sheetCornerRadius={-1}
      sheetExpandsWhenScrolledToEdge={true}
      sheetGrabberVisible={false}
      sheetLargestUndimmedDetent="all"
      stackPresentation="push"
      style={
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      swipeDirection="horizontal"
    >
      <View
        accessibilityElementsHidden={false}
        importantForAccessibility="auto"
        style={
          {
            "flex": 1,
            "flexDirection": "column-reverse",
          }
        }
      >
        <View
          collapsable={false}
          style={
            [
              {
                "flex": 1,
              },
              {
                "backgroundColor": "white",
              },
              undefined,
            ]
          }
        >
          <View
            style={
              {
                "backgroundColor": "#ffffff",
                "flex": 1,
              }
            }
          >
            <Text
              testID="user.test.username"
            >
              yssk22
            </Text>
          </View>
        </View>
      </View>
      <RNSScreenStackHeaderConfig
        backButtonInCustomView={false}
        backTitle=""
        backTitleVisible={false}
        backgroundColor="#0075c2"
        color="white"
        direction="ltr"
        disableBackButtonMenu={false}
        hidden={false}
        hideBackButton={false}
        largeTitleHideShadow={false}
        title="/"
        titleColor="white"
        topInsetEnabled={false}
        translucent={false}
      />
    </RNSScreen>
  </RNSScreenStack>
</View>
`;

exports[`user useHelloProject 1`] = `
<View
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <RNSScreenStack
    style={
      {
        "flex": 1,
      }
    }
  >
    <RNSScreen
      collapsable={false}
      fullScreenSwipeShadowEnabled={false}
      gestureResponseDistance={
        {
          "bottom": -1,
          "end": -1,
          "start": -1,
          "top": -1,
        }
      }
      hasLargeHeader={false}
      nativeBackButtonDismissalEnabled={false}
      onAppear={[Function]}
      onDisappear={[Function]}
      onDismissed={[Function]}
      onGestureCancel={[Function]}
      onHeaderBackButtonClicked={[Function]}
      onNativeDismissCancelled={[Function]}
      onTransitionProgress={[Function]}
      onWillAppear={[Function]}
      onWillDisappear={[Function]}
      replaceAnimation="push"
      sheetAllowedDetents="large"
      sheetCornerRadius={-1}
      sheetExpandsWhenScrolledToEdge={true}
      sheetGrabberVisible={false}
      sheetLargestUndimmedDetent="all"
      stackPresentation="push"
      style={
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      swipeDirection="horizontal"
    >
      <View
        accessibilityElementsHidden={false}
        importantForAccessibility="auto"
        style={
          {
            "flex": 1,
            "flexDirection": "column-reverse",
          }
        }
      >
        <View
          collapsable={false}
          style={
            [
              {
                "flex": 1,
              },
              {
                "backgroundColor": "white",
              },
              undefined,
            ]
          }
        >
          <View
            style={
              {
                "backgroundColor": "#ffffff",
                "flex": 1,
              }
            }
          >
            <Text
              testID="user.test.useArtist"
            >
              morningmusume
            </Text>
            <Text
              testID="user.test.useMember"
            >
              mizuki_fukumura
            </Text>
            <Text
              testID="user.test.useArtists"
            >
              morningmusume
angerme
juicejuice
tsubakifactory
beyooooonds
ochanorma
rosychronicle
            </Text>
            <Text
              testID="user.test.useMembers"
            >
              erina_ikuta
ayumi_ishida
sakura_oda
miki_nonaka
maria_makino
akane_haga
reina_yokoyama
rio_kitagawa
homare_okamura
mei_yamazaki
rio_sakurai
haruka_inoue
ako_yumigeta
ayano_kawamura
moe_kamikokuryo
layla_ise
rin_hashisako
rin_kawana
shion_tamenaga
wakana_matsumoto
yuki_hirayama
yukiho_shimoitani
hana_goto
ruru_dambara
yume_kudou
riai_matsunaga
rei_inoue
ichika_arisawa
risa_irie
kisaki_ebata
sakura_ishiyama
akari_endo
mifu_kawashima
ami_tanimoto
mizuho_ono
saori_onoda
mao_akiyama
yuumi_kasai
shiori_yagi
marine_fukuda
runo_yofu
mihane_ishii
yuu_murata
fuka_doi
rika_shimakura
shiori_nishida
saya_eguchi
kurumi_takase
kokoro_maeda
minami_okamura
momohime_kiyono
miyo_hirai
honoka_kobayashi
utano_satoyoshi
madoka_saito
ruli_hiromoto
kanami_ishiguri
kirara_yonemura
nanami_kubota
sumire_tashiro
natsume_nakayama
miku_nishizaki
momo_kitahara
roko_tsutsui
honoka_hashida
hinoha_yoshida
karin_onoda
ayana_murakoshi
hasumi_uemura
yulia_matsubara
hana_shimakawa
rena_kamimura
yume_soma
            </Text>
          </View>
        </View>
      </View>
      <RNSScreenStackHeaderConfig
        backButtonInCustomView={false}
        backTitle=""
        backTitleVisible={false}
        backgroundColor="#0075c2"
        color="white"
        direction="ltr"
        disableBackButtonMenu={false}
        hidden={false}
        hideBackButton={false}
        largeTitleHideShadow={false}
        title="/"
        titleColor="white"
        topInsetEnabled={false}
        translucent={false}
      />
    </RNSScreen>
  </RNSScreenStack>
</View>
`;
